indexing-service/src/main/java/org/apache/druid/indexing/firehose/IngestSegmentFirehoseFactory.java

```diff
@@ -251,9 +273,169 @@ private long jitter(long input)
+    final Map<DataSegment, WindowedSegment> windowedSegmentsBySegment = new HashMap<>();
+
+    for (TimelineObjectHolder<String, DataSegment> timelineHolder : timeLineSegments) {
+      for (PartitionChunk<DataSegment> chunk : timelineHolder.getObject()) {
+        final DataSegment segment = chunk.getObject();
+        if (timelineHolder.getInterval().isEqual(timelineHolder.getTrueInterval())) {
+          // This is the simple case: the entirety of the segment is used: no overshadowing or clipping
+          // from the beginning or end of the full interval. We aren't going to see it again.
+          if (windowedSegmentsBySegment.containsKey(segment)) {
+            throw new ISE("unclipped segment[%s] seen more than once on timeline", segment);
+          }
+          final WindowedSegment windowedSegment = new WindowedSegment(segment, null);
+          windowedSegments.add(windowedSegment);
+          windowedSegmentsBySegment.put(segment, windowedSegment);
+        } else {
+          // Some part of the segment is overshadowed or clipped. This WindowedSegment will need
+          // to have explicit intervals.
+          final WindowedSegment existingWindowedSegment = windowedSegmentsBySegment.get(segment);
+          if (existingWindowedSegment == null) {
+            final List<Interval> intervals = new ArrayList<>();
+            intervals.add(timelineHolder.getInterval());
+            final WindowedSegment newWindowedSegment = new WindowedSegment(segment, intervals);
+            windowedSegments.add(newWindowedSegment);
+            windowedSegmentsBySegment.put(segment, newWindowedSegment);
+          } else {
+            if (!existingWindowedSegment.hasExplicitIntervals()) {
+              throw new ISE("unclipped segment[%s] seen later clipped on timeline", segment);
+            }
+            existingWindowedSegment.getIntervals().add(timelineHolder.getInterval());
+          }
+        }
+      }
+    }
+
+    return windowedSegments;
+  }
+
+  private List<WindowedSegment> getWindowedSegmentsForIds(List<WindowedSegmentId> ids)
+  {
+    // FIXME This is doing a series of single HTTP calls for each segment. It could instead do a single
+    // GET /metadata/datasources/DATASOURCE/segments call to download all DataSegments. I'm erring on the
+    // side of avoiding single unboundedly-large calls, in favor of more smaller roundtrip --- after all,
+    // we're going to later fetch each of these segments in series, so the additional coordinator calls
+    // in series here don't feel that bad. And adding a new "get multiple segments by ID" HTTP call seems
+    // like overkill.
+    //
+    // Also: If the only use of WindowedSegments and the "segments" field on this FirehoseFactory is for internal
+    // use by parallel ingestion, then it might make sense to get rid of WindowedSegmentId entirely and just put
+    // WindowedSegments directly into this class's 'segments' parameter. Then we wouldn't even need to do this fetch
```

Comment:
Another concern is that, Druid is using ZooKeeper for assigning tasks which means the overlord writes task specs in some ZooKeeper nodes to assign them. `SegmentIds` is safer than `DataSegments` because there's a limit for max bytes of data written in a node.